source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h

```diff
@@ -21,16 +22,31 @@ class LocalRateLimiterImpl {
 
   bool requestAllowed() const;
 
+  void fillDescriptor(const std::vector<Envoy::RateLimit::LocalDescriptor>& descriptors);
+
+  bool requestAllowed(const Envoy::RateLimit::LocalDescriptor& descriptor) const;
+
 private:
+  struct Token {
+    mutable std::atomic<uint32_t> tokens;
+  };
   void onFillTimer();
+  void onFillTimerHelper(const Token& tokens, const uint32_t max_tokens,
+                         const uint32_t tokens_per_fill);
+  void onFillTimerDescriptorHelper();
+  bool requestAllowedHelper(const Token& tokens) const;
 
   const std::chrono::milliseconds fill_interval_;
   const uint32_t max_tokens_;
   const uint32_t tokens_per_fill_;
   const Event::TimerPtr fill_timer_;
-  mutable std::atomic<uint32_t> tokens_;
+  Token tokens_;
   mutable Thread::ThreadSynchronizer synchronizer_; // Used for testing only.
 
+  absl::flat_hash_map<Envoy::RateLimit::LocalDescriptor, std::unique_ptr<Token>>
+      tokens_per_descriptor_;
+  absl::flat_hash_map<Envoy::RateLimit::LocalDescriptor, long int> last_refill_per_descriptor_;
```

Comment:
This can be uint64_t or better just std::chrono::milliseconds if you are actually storing a time in here (or some other monotonic time point type)

Reply:
using std::chrono::time_point